src/utils.rs

```diff
@@ -17,3 +19,145 @@ pub(crate) fn mlx_describe(ptr: *mut ::std::os::raw::c_void) -> Option<String> {
 
     description
 }
+
+/// Helper method to check if two arrays are broadcastable.
+///
+/// Uses the same broadcasting rules as numpy.
+/// https://numpy.org/doc/1.20/user/theory.broadcasting.html
+///
+/// "The size of the trailing axes for both arrays in an operation must
+/// either be the same size or one of them must be one."
+pub(crate) fn is_broadcastable(a: &Array, b: &Array) -> bool {
+    let shape_a = a.shape();
+    let shape_b = b.shape();
+
+    shape_a
+        .iter()
+        .rev()
+        .zip(shape_b.iter().rev())
+        .all(|(a, b)| *a == 1 || *b == 1 || a == b)
+}
+
+impl Array {
+    /// Helper method to check if an array can be reshaped to a given shape.
+    pub fn can_reshape_to(&self, shape: &[i32]) -> bool {
+        if self.shape() == shape {
+            return true;
+        }
+
+        let mut size = 1;
+        let mut infer_idx: isize = -1;
+        for i in 0..shape.len() {
```

Comment:
Just curious when should a shape carry negative element

Reply:
good question, just ported this over from mlx

Reply:
I wonder what is negative shape intended for? Is -1 the only possible negative value?

Reply:
resolved on discord + added one more test for dimensions other than -1